codalab/model/bundle_model.py

```diff
@@ -699,6 +699,78 @@ def get_bundle_uuids(self, conditions, max_results):
 
         return self._execute_query(query)
 
+    def get_memoized_bundles(self, user_id, command, dependencies):
+        '''
+        Get a list of bundle UUIDs that match with input command and dependencies in the order they were created.
+        :param user_id: a string that specifies the current user id.
+        :param command: a string that defines the command that is used to search for memoized bundles in the database.
+        :param dependencies: a string in the form of '[{"child_path": key1, "parent_uuid": uuid1},
+                                                       {"child_path": key2, "parent_uuid": uuid2}]'
+                             to search for memoized bundles in the database.
+        :return: a list of matched UUIDs in the order they were created.
+        '''
+        # Decode json formatted dependencies string to a list of key value pairs
+        dependencies = json.loads(dependencies)
+        # When there is no dependency to be matched, the target memozied bundle
+        # should only exist in the bundle table but not in the bundle_dependency table.
+        if len(dependencies) == 0:
+            query = (
+                select([cl_bundle.c.uuid])
+                .select_from(cl_bundle)
+                .where(
+                    and_(
+                        cl_bundle.c.command == command,
+                        cl_bundle.c.owner_id == user_id,
+                        cl_bundle.c.uuid.notin_(
+                            select([cl_bundle_dependency.c.child_uuid]).select_from(
+                                cl_bundle_dependency
+                            )
+                        ),
+                    )
+                )
+                .order_by(cl_bundle.c.id)
+            )
+        else:
+            clause = []
```

Comment:
Please provide a more conceptual, high-level overview of this logic.

Reply:
I should have provided this when I created this PR. I have added a section in [here](https://github.com/codalab/codalab-worksheets/pull/2159#issue-403439363) to describe what the high-level logic is. Please let me know if this is clear.

Reply:
I should have created one at the beginning when created this PR. I added more details on https://github.com/codalab/codalab-worksheets/pull/2159#issue-403439363.

Reply:
Could you please add a shortened version of this into the code itself?  It will be hard to find this PR if you're just browsing the code.

Reply:
I have added a shortened version of logic into this code. Also I have updated the google doc with more information [new memoize option](https://docs.google.com/document/d/1UDjagq17qbVdxfAPJJJA6qyqArfnyyLt_wAQQpOVzIY/edit)